File: AlgorithmsProblemSolvingJava/src/io/programminglife/algoexpert/easy/GenerateDocument.java
package io.programminglife.algoexpert.easy;

import java.util.HashMap;
import java.util.Map;

public class GenerateDocument {

    public static boolean generateDocument(String characters, String document) {
        Map<Character, Integer> charFreq = new HashMap<Character, Integer>();

        for (int i = 0; i < characters.length(); i++) {
            Character current = characters.charAt(i);
            charFreq.put(current, charFreq.getOrDefault(current, 0) + 1);
        }

        for (int i = 0; i < document.length(); i++) {
            Character current = document.charAt(i);
            if (!charFreq.containsKey(current) || charFreq.get(current) == 0) {
                return false;
            }
            charFreq.put(current, charFreq.get(current) - 1);
        }

        return true;
    }

    public static void main(String[] args) {
        String characters = "Bste!hetsi ogEAxpelrt x ";
        String document = "AlgoExpert is the Best!";

        System.out.println(generateDocument(characters, document));
    }

}

File: AlgorithmsProblemSolvingPython/algoexpert/easy/generateDocument.py
def generateDocument(characters, document):
    charCount = {}

    for character in characters:
        if character not in charCount:
            charCount[character] = 0
        charCount[character] += 1

    for character in document:
        if character not in charCount or charCount[character] == 0:
            return False
        charCount[character] -= 1

    return True


if __name__ == "__main__":
    characters = "Bste!hetsi ogEAxpelrt x "
    document = "AlgoExpert is the Best!"

    print(generateDocument(characters, document))
